src/git-repository-async.js

```diff
@@ -0,0 +1,873 @@
+      .catch(_ => null)
+  }
+
+  // Public: Get the URL for the 'origin' remote.
+  //
+  // * `path` (optional) {String} path in the repository to get this information
+  //   for, only needed if the repository has submodules.
+  //
+  // Returns a {Promise} which resolves to the {String} origin url of the
+  // repository.
+  getOriginURL (_path) {
+    return this.getConfigValue('remote.origin.url', _path)
+  }
+
+  // Public: Returns the upstream branch for the current HEAD, or null if there
+  // is no upstream branch for the current HEAD.
+  //
+  // * `path` An optional {String} path in the repo to get this information for,
+  //   only needed if the repository contains submodules.
+  //
+  // Returns a {Promise} which resolves to a {String} branch name such as
+  // `refs/remotes/origin/master`.
+  getUpstreamBranch (_path) {
+    return this.getRepo(_path)
+      .then(repo => repo.getCurrentBranch())
+      .then(branch => Git.Branch.upstream(branch))
+  }
+
+  // Public: Gets all the local and remote references.
+  //
+  // * `path` An optional {String} path in the repository to get this information
+  //   for, only needed if the repository has submodules.
+  //
+  // Returns a {Promise} which resolves to an {Object} with the following keys:
+  //  * `heads`   An {Array} of head reference names.
+  //  * `remotes` An {Array} of remote reference names.
+  //  * `tags`    An {Array} of tag reference names.
+  getReferences (_path) {
+    return this.getRepo(_path)
+      .then(repo => repo.getReferences(Git.Reference.TYPE.LISTALL))
+      .then(refs => {
+        const heads = []
+        const remotes = []
+        const tags = []
+        for (const ref of refs) {
+          if (ref.isTag()) {
+            tags.push(ref.name())
+          } else if (ref.isRemote()) {
+            remotes.push(ref.name())
+          } else if (ref.isBranch()) {
+            heads.push(ref.name())
+          }
+        }
+        return {heads, remotes, tags}
+      })
+  }
+
+  // Public: Get the SHA for the given reference.
+  //
+  // * `reference` The {String} reference to get the target of.
+  // * `path` An optional {String} path in the repo to get the reference target
+  //   for. Only needed if the repository contains submodules.
+  //
+  // Returns a {Promise} which resolves to the current {String} SHA for the
+  // given reference.
+  getReferenceTarget (reference, _path) {
+    return this.getRepo(_path)
+      .then(repo => Git.Reference.nameToId(repo, reference))
+      .then(oid => oid.tostrS())
+  }
+
+  // Reading Status
+  // ==============
+
+  // Public: Resolves true if the given path is modified.
+  //
+  // * `path` The {String} path to check.
+  //
+  // Returns a {Promise} which resolves to a {Boolean} that's true if the `path`
+  // is modified.
+  isPathModified (_path) {
+    return this._filterStatusesByPath(_path).then(statuses => {
+      return statuses.filter(status => status.isModified()).length > 0
```

Comment:
Can we use [Array.prototype.some](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/some)
```javascript
return statuses.some(status => status.isModified())
```
here?

The thing is that it can return earlier without iterating whole collection and also not producing new array.